common/src/static_cell.rs

```diff
@@ -0,0 +1,76 @@
+#[cfg(not(feature = "threading"))]
+mod non_threading {
+    use crate::lock::OnceCell;
+    pub type StaticCell<T> = std::thread::LocalKey<OnceCell<&'static T>>;
+
+    #[macro_export]
+    macro_rules! static_cells {
+        // process multiple declarations
+        ($(#[$attr:meta])* $vis:vis static $name:ident: $t:ty;) => (
+            std::thread_local! {
+                $(#[$attr])* $vis static $name: $crate::lock::OnceCell<&'static $t> = $crate::lock::OnceCell::new();
+            }
+        );
+    }
+}
+#[cfg(not(feature = "threading"))]
+pub use non_threading::*;
+
+#[cfg(feature = "threading")]
+mod threading {
+    use crate::lock::OnceCell;
+
+    pub struct StaticKey<T> {
+        inner: T,
+    }
+
+    impl<T> StaticKey<T> {
+        pub const fn new(inner: T) -> Self {
+            Self { inner }
+        }
+
+        pub fn with<F, R>(&self, f: F) -> R
+        where
+            F: FnOnce(&T) -> R,
+        {
+            f(&self.inner)
+        }
+    }
+
+    pub type StaticCell<T> = StaticKey<OnceCell<&'static T>>;
```

Comment:
I think you should be able to make this `OnceCell<T>` -- try modeling the non-threading version after the OnceCell API, rather than modeling the threading version off the LocalKey API, since OnceCell can provide a `&'static T` by itself, but LocalKey should still need the `Box::leak()` stuff.

Reply:
Yes, that should be possible. But this is easier by providing exact same API with `LocalKey`.
At the same time, the reference of immutable static variable is practically just same as permanently leaked reference.

So I agree that I was able to do it, but still don't have any reason to do it (yet).

Let me know if i am missing something.